### PR TITLE
Add param to enable/disable saving ONNX model as external data

### LINKED ISSUE #8522: [FR] Parametrize the save as external data for ONNX model logging
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

Since #7808, ONNX models are systematically saved with the weights saved in a separate file. This introduced a behavior change, and it is needed only for big models. The associated code is here:

https://github.com/mlflow/mlflow/blob/5712024236c45e7830506bb3adba3c6e42131625/mlflow/onnx.py#L175

It would be nice to allow to enable or not this behavior so that a small model can still be serialized as a single file.

### Motivation

> #### What is the use case for this feature?

- Keep the behavior consistent this the previous one
- Serialize small models as a single file

> #### Why is this use case valuable to support for MLflow users in general?

Splitting a model into several files makes using it more complex, primarily when mlflow is not used for inference.

> #### Why is this use case valuable to support for your project(s) or organization?

We never use models bigger than 2GB, and we process models are single-file objects.

> #### Why is it currently difficult to achieve this use case?

It is not possible to log ONNX models as a single file with recent releases of mlflow.

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by daniellok-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/daniellok-db/mlflow/pull/10152?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10152/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10152
```

</p>
</details>

### Related Issues/PRs

Resolves #8522

### What changes are proposed in this pull request?

WIP for now, I can't seem to get the ONNX tests to run locally so I'm creating a PR to see if the CI runner has better luck

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

ONNX model saving API now accepts a `save_as_external_data` param (defaults to `True`). If set to `False`, it will not save tensors in a separate file.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by daniellok-db** [comment]
nevermind the tests run they just take super long

File: tests/onnx/test_onnx_model_export.py [hunk 1]
```diff
@@ -253,16 +253,11 @@ def test_model_save_load(onnx_model, model_path):
...
+        onnx.convert_model_to_external_data = mock.Mock()
 
-        def onnx_save_nonexternal(
-            model, path, save_as_external_data
-        ):  # pylint: disable=unused-argument
-            original_save_model(model, path, save_as_external_data=False)
-
-        with mock.patch("onnx.save_model", wraps=onnx_save_nonexternal):
-            mlflow.onnx.save_model(onnx_model, model_path)
+        mlflow.onnx.save_model(onnx_model, model_path, save_as_external_data=False)
```
Comment (BenWilson2) [inline tests/onnx/test_onnx_model_export.py:259]: Can we parametrize one (just one of them, no need to do any more than that) of the tests below that actually save and load an ONNX model with this option set explicitly to True and False just to be certain that we're able to create inference-capable models with this argument setting?

File: tests/onnx/test_onnx_model_export.py [hunk 2]
```diff
@@ -346,10 +341,17 @@ def test_model_save_load_multiple_inputs(onnx_model_multiple_inputs_float64, mod
     assert onnx.checker.check_model.called
 
 
+@pytest.mark.parametrize("save_as_external_data", [True, False])
```
Comment (daniellok-db) [inline tests/onnx/test_onnx_model_export.py:344]: @BenWilson2 will this do?
Screenshot 2023-10-27 at 8 49 27 AM

Comment (BenWilson2) [reply tests/onnx/test_onnx_model_export.py:344]: Perfect! :)

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks for the regression fix here @daniellok-db :)

File: mlflow/onnx/__init__.py [hunk 3]
```diff
@@ -519,4 +523,5 @@ def log_model(
         onnx_execution_providers=onnx_execution_providers,
         onnx_session_options=onnx_session_options,
         metadata=metadata,
+        save_as_external_data=save_as_external_data
```
Comment (daniellok-db) [inline mlflow/onnx/__init__.py:526]: it looks like the `**kwargs` get passed through to the `save_model` function here:
https://github.com/mlflow/mlflow/blob/master/mlflow/models/model.py#L600

Comment (BenWilson2) [reply mlflow/onnx/__init__.py:526]: yep! 100%

**Pull request merged**